### ISSUE TITLE
Should I apply softmax or log_softmax to my token_scores?

### ISSUE THREAD

**Issue by poteminr** [kmkurn/pytorch-crf#108]
```
embedded_text_input = self.encoder(input_ids=input_ids, attention_mask=attention_mask).last_hidden_state
embedded_text_input = self.dropout(F.leaky_relu(embedded_text_input))
 token_scores = F.log_softmax(self.feedforward(embedded_text_input), dim=-1)
 # or
 # token_scores = self.feedforward(embedded_text_input)

loss, output_tags = self.apply_crf(token_scores, labels, attention_mask, batch_size=batch_size)
```
Should I apply softmax before passing token_scores to the CRF?

**Comment by kmkurn** [comment]
No you don’t have to. The CRF is already a (very large) softmax over the possible tag sequences.On 2 Dec 2022, at 4:42 am, Roman Potemin ***@***.***> wrote:﻿
embedded_text_input = self.encoder(input_ids=input_ids, attention_mask=attention_mask).last_hidden_state
embedded_text_input = self.dropout(F.leaky_relu(embedded_text_input))
 token_scores = F.log_softmax(self.feedforward(embedded_text_input), dim=-1)
 # or
 # token_scores = self.feedforward(embedded_text_input)

loss, output_tags = self.apply_crf(token_scores, labels, attention_mask, batch_size=batch_size)

Should I apply softmax before passing token_scores to the CRF?

**Comment by poteminr** [comment]
Thank you!
